internal/util/util.go

```diff
@@ -80,3 +82,165 @@ func removeNullValuesFromJSONMap(m map[string]interface{}) bool {
 	}
 	return false
 }
+
+// SplitObject returns 2 maps from a given object as bytes array and a slice of paths.
+// One of the 2 output maps will include only the paths from the slice while the second returned map will include all other paths.
+func SplitObject(obj []byte, paths []string, pathSeparator string) (map[string]any, map[string]any, error) {
+	var objAsMap map[string]any
+	if err := json.Unmarshal(obj, &objAsMap); err != nil {
+		return nil, nil, err
+	}
+
+	return SplitMap(objAsMap, paths, pathSeparator)
+}
+
+// SplitMap returns 2 maps from a given map and a slice of paths.
+// One of the 2 output maps will include only the paths from the slice while the second returned map will include all other paths.
+func SplitMap(obj map[string]any, paths []string, pathSeparator string) (onlyPaths map[string]any, withoutPaths map[string]any, err error) {
+	onlyPaths = make(map[string]any)
+	withoutPaths, err = clone(obj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// In case there are no paths, return an empty map and the original map.
+	// This is in place so that this logic can be bypassed at runtime by providing
+	// an empty slice of paths.
+	if len(paths) == 0 {
+		return onlyPaths, withoutPaths, nil
+	}
+
+	// Sort the paths slice before using it to avoid maps merging in case the slice were to
+	// include deeply nested fields and then the higher level fields (ex: []string{"map.field.inner2", "map"})
+	sort.Strings(paths)
+
+	for _, path := range paths {
+		pathTokens := strings.Split(path, pathSeparator)
+
+		if err := extractPath(withoutPaths, onlyPaths, pathTokens); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	cleanup(onlyPaths)
+	cleanup(withoutPaths)
+
+	return onlyPaths, withoutPaths, nil
+}
+
+// clone returns a clone of the given map
+func clone(obj map[string]any) (map[string]any, error) {
+	var clone map[string]any
+
+	if err := StructToStruct(obj, &clone); err != nil {
+		return nil, err
+	}
+
+	return clone, nil
+}
+
+// cleanup removes nil values, empty maps, and empty arrays from the given object
+func cleanup(obj map[string]any) {
+	for key, val := range obj {
+		switch typedVal := val.(type) {
+		case map[string]any:
+			cleanup(typedVal)
+
+			if len(typedVal) == 0 {
+				delete(obj, key)
+			}
+
+		case []any:
+			for i := 0; i < len(typedVal); i++ {
+				if elemMap, ok := typedVal[i].(map[string]any); ok {
+					cleanup(elemMap)
+
+					if len(elemMap) == 0 {
+						typedVal = append(typedVal[:i], typedVal[i+1:]...)
+						i--
+					}
+				} else if typedVal[i] == nil {
+					typedVal = append(typedVal[:i], typedVal[i+1:]...)
+					i--
+				}
+			}
+
+			if len(typedVal) == 0 {
+				delete(obj, key)
+			} else {
+				obj[key] = typedVal
+			}
+
+		case nil:
+			delete(obj, key)
+
+		}
```

Comment:
will you need to incorporate my recent [fix](https://github.com/numaproj/numaplane/pull/213) to look for keys with zero values?

Reply:
Yes, I will add those cases since they are needed. Alternatively, I could use your `removeNullValuesFromJSONMap` instead of the `cleanup` function. What do you prefer?

Reply:
whatever you prefer. I guess the name `cleanup()` should probably at least be modified to be more clear about what it means